tictactoe.py

```diff
@@ -0,0 +1,162 @@
+    
+    no_of_x, no_of_o = 0, 0
+    for row in board:
+        for position in row:
+            if position == X:
+                no_of_x += 1
+            elif position == O:
+                no_of_o += 1
+    if start == X:
+        if no_of_x > no_of_o:
+            return O
+        else:
+            return X
+    elif start == O:
+        if no_of_x < no_of_o:
+            return X
+        else:
+            return O
+"""
+
+
+def update_board(main_embed, reaction, player):
+    board[reaction] = player
+    main_embed.description = update_board_embed()
+    return main_embed
+
+
+def update_board_embed():
+    board_result = ""
+    for i in range(3):
+        line1 = ""
+        line2 = ""
+        for j in ["↖", "⬆", "↗", "⬅", "⏺", "➡", "↙", "⬇", "↘"][i * 3:(i + 1) * 3]:
+            line1 += ''.join(board[j][0:2])
+            line2 += ''.join(board[j][2:4])
+        board_result += line1 + "\n" + line2 + "\n"
+    return board_result
+
+def reset_board():
+    global board
+    board = {"↖": EMPTY, "⬆": EMPTY, "↗": EMPTY,
+             "⬅": EMPTY, "⏺": EMPTY, "➡": EMPTY,
+             "↙": EMPTY, "⬇": EMPTY, "↘": EMPTY}
+
+def actions(board):
+    """
+    Returns set of all possible actions (i, j) available on the board.
+    """
+
+    action = list()
+    for row, i in enumerate(board):
+        for position, j in enumerate(i):
+            if j is None:
+                val = row, position
+                action.append(val)
+    return action
+
+
+def valid_action(move: tuple, board: list) -> bool:
+    return move in actions(board)
+
+"""
+def result(board, action):
+    
+    Returns the board that results from making move (i, j) on the board.
+    
+    new_board = list(board)
+    try:
+        if new_board[action[0]][action[1]] is None:
+            new_board[action[0]][action[1]] = player(board)
+        else:
+            raise ValueError("Already filled position.")
+    except:
+        raise ValueError("Position out of board.")
+    return new_board
+"""
+
+
+def winner(board):
+    """
+    Returns the winner of the game, if there is one.
+    """
+    # if board is None:
+    current_board = initial_state()
+    for i in range(3):
+        if current_board[i*3] == current_board[i*3+1] == current_board[i*3+2] != EMPTY:
+            return True
+    for i in range(3):
+        if current_board[i] == current_board[i+3] == current_board[i+6] != EMPTY:
+            return True
+    if current_board[0] == current_board[4] == current_board[8] != EMPTY:
+        return True
+    if current_board[2] == current_board[4] == current_board[6] != EMPTY:
+        return True
+    return False
+
+
+
+
+
+def terminal(board):
+    """
+    Returns True if game is over, False otherwise.
+    """
+    if winner(board) is True:
+        reset_board()
+        return True
+    else:
+        for i in board:
+            if i == EMPTY:
+                return False
+
+
+def utility(board):
```

Comment:
Are we planning on finishing a minimax algorithm to have an option to play against an AI? 
Would be fun I think, if not we can remove these two functions from the code

Reply:
We can develop the minimax later I think, there are other things to work on in the bot right?
We can just comment the functions out.